### PR TITLE
explain: deprecate and remove the old `explain` code

### PR DESCRIPTION
**Pull request by aalexandrov**
This is a preparation commit for #17360.

### Motivation

  * This PR adds a known-desirable feature.

Part of #17360.

### Tips for reviewer

What I managed to achieve in this PR:

1. Remove two of the three `explain.rs` files (handling rendering of `Hir~Expr` and `DataflowDesc`.
2. Deprecate the remaining `explain.rs` files and mark the remaining bits with `#[allow(deprecated)] // TODO(#17360)`.

We should agree on a plan to remove (2) in the umbrella issue (#17360).

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/compute-client/src/explain.rs [hunk 1]
```diff
@@ -27,14 +27,14 @@
 //! printed in contexts where trailing whitespace is unacceptable, like
 //! sqllogictest files.
 
+#![deprecated = "Use `explain_new` instead."]
+
 use std::fmt;
```
Comment (ggevay) [inline src/compute-client/src/explain.rs:33]: The `pub trait ViewFormatter` in this file has no implementations according to my IDE. Maybe you intended to delete this file?

You also wrote
> Remove two of the three explain.rs files

but there seem to be 2 `explain.rs` files remaining.

Comment (aalexandrov) [reply src/compute-client/src/explain.rs:33]: I forgot to remove the one in `compute-client`.

File: src/repr/src/explain_new/mod.rs [hunk 2]
```diff
@@ -666,6 +666,36 @@ where
     Ok(())
 }
 
+/// Pretty-prints a list of indices.
+#[derive(Debug)]
+pub struct Indices<'a>(pub &'a [usize]);
```
Comment (ggevay) [inline src/repr/src/explain_new/mod.rs:671]: I think the old copy of this in (the remaining) `explain.rs` could be removed. `explain.rs` could just import the new copy.

**Review by ggevay — APPROVED** [review]

File: src/repr/src/explain_new/mod.rs (hunk 2 above)
Comment (aalexandrov) [reply src/repr/src/explain_new/mod.rs:671]: I wanted to avoid this work as very briefly I will wipe `explain.rs` altogether.

**Pull request merged**